### ISSUE TITLE
[GUIDE] How to install Betterfox on Android Firefox using adb

### ISSUE THREAD

**Issue by gunir** [yokoffing/Betterfox#240]
1. Install adb: https://www.xda-developers.com/install-adb-windows-macos-linux/

2. Connect your Android device to your PC using cable, if success you'll be able to see something like this if you type `adb devices` and Enter:
![image](https://github.com/yokoffing/Betterfox/assets/134402102/3407554d-145b-4411-9fbc-4e3f6756c7ab)

3. Enable `Remote Debugging via USB` in Android Firefox's Settings:
![image](https://github.com/yokoffing/Betterfox/assets/134402102/4c113a46-fd1e-48f3-a90f-4fe328ecae05)

4. From Firefox PC, type `about:debugging` and Enter, you'll see your Android device in the list, choose Connect:
![image](https://github.com/yokoffing/Betterfox/assets/134402102/41696630-531d-479e-8072-2713623a61e4)

5. And open your device and you'll see all the opening tabs:
![image](https://github.com/yokoffing/Betterfox/assets/134402102/a6cfcbe7-bfed-4b78-bc3d-15bf96523b71)

6. Open `about:support` in your Firefox Android
7. Inspect the `about:support` tab from your Firefox PC, you will see this console screen:
![image](https://github.com/yokoffing/Betterfox/assets/134402102/6746b9f4-5591-4806-a9cb-aae470294ae6)

8. Paste this code to the console:
```


var user_pref = function(pref, val){

  try {
    if(typeof val == "string"){

         Services.prefs.setStringPref(pref, val);    
    }
    else if(typeof val == "number"){

         Services.prefs.setIntPref(pref, val);    
    }
    else if(typeof val == "boolean"){

         Services.prefs.setBoolPref(pref, val);    
    }
  } catch(e){
    console.log("pref:" + pref + " val:" + val + " e:" + e);
  }
}
//paste your user.js file content here


```

10. Now the final step, copy Betterfox's code, for example [Fasterfox](https://raw.githubusercontent.com/yokoffing/Betterfox/main/Fastfox.js) and replace `//paste your user.js file content here` then Enter.

You're done, now all the Fasterfox is applied to your Android Firefox.

This guide was written by a Vietnamese Firefox user with nickname @ndv92, this is the original post: https://voz.vn/t/tong-hop-nhung-addon-chat-cho-firefox-pc-mobile.682181/post-27739740

This Vietnamese thread is surprisingly deep about Firefox.

None was written by me, I just translate it to English, literally.

Thanks for reading,
And thank @yokoffing for making and maintaining Betterfox.

**Comment by vonwerderc** [comment]
@gunir Do we check about:config in the browser or the phone? The tab will not load when I check it on my phone, so I can't really tell if it worked. I was able to do all of the steps though.

**Comment by gunir** [comment]
> @gunir Do we check about:config in the browser or the phone? The tab will not load when I check it on my phone, so I can't really tell if it worked. I was able to do all of the steps though.

Yes, but you need to use Firefox Beta/Nightly, Stable doesn't allow to open it for some reasons ?

**Comment by yokoffing** [comment]
Thanks @gunir for posting and translating this mini-guide. Have you tried it yourself? Does it work?

**Comment by vonwerderc** [comment]
> Thanks @gunir for posting and translating this mini-guide. Have you tried it yourself? Does it work?

It worked for me using Firefox beta on Android. It might work on regular firefox, but you can't check about:config to verify.

**Comment by THEBOSS619** [comment]
Just wanted to say, it is confirmed and working on Lastest Firefox Beta on Android today. All good!

**Comment by kuriokurio** [comment]
hey code newbie here, I recently managed to set up hardened firefox on my pc ^-^ just ran through this post today and decided to setup hardened firefox on my  android phone as well, I followed the steps ans used securefox.js on nightly. user above confirmed it was working, I'm not advanced enough to test that but it seems mine worked too ! thanks to all

**Comment by xd003** [comment]
Can we please have this mentioned in the readme somewhere, i wish i would have find this earlier.
 Thanks for the guide

**Comment by gunir** [comment]
> Can we please have this mentioned in the readme somewhere, i wish i would have find this earlier. Thanks for sharing the guide

How about writing a wiki post ?

https://github.com/yokoffing/Betterfox/wiki

**Comment by xd003** [comment]
> > Can we please have this mentioned in the readme somewhere, i wish i would have find this earlier. Thanks for sharing the guide
> 
> How about writing a wiki post ?
> 
> https://github.com/yokoffing/Betterfox/wiki

Will work but i was thinking we could rather a have a Betterfox user.js specifically for Android itself. Since not all the flags from desktop are present on Android, moreover there could be certain Android specific flags which could be benefited from.

**Comment by gunir** [comment]
> Since not all the flags from desktop are present on Android, moreover there could be certain Android specific flags which could be benefited from.

So far so good from what I've tested, using the base user.js on Firefox Android only improves speed and battery life, which is great because telemetry can be very battery hungry.

I've seen no issues.

**Comment by Zachary-Luckman** [comment]
Still works in 2024 on Firefox Beta. Hope this gets picked up by the dev and turns official

**Comment by tailwhipstruggle** [comment]
I can't get this to work in Firefox beta, dunno if something changed recently but for every line in the user.js it outputs "Services is not defined". I imagine a simple edit to the script could make it work again but I don't know enough about Firefox to find out what to change

**Comment by gunir** [comment]
> I can't get this to work in Firefox beta, dunno if something changed recently but for every line in the user.js it outputs "Services is not defined". I imagine a simple edit to the script could make it work again but I don't know enough about Firefox to find out what to change
> 
> Edit: same with firefox nightly

Make sure you run it on `about:support` page, just need to do exactly as the guide, it's pretty much the most informative guide already, nothing much can be improved.

**Comment by iHarryPotter178** [comment]
Fennec from Fdroid allows about:config and also it's based on Firefox stable, not beta or nightly. Update comes a week or so later than Firefox. But this could be used for this? I haven't tested yet..

**Comment by gunir** [comment]
> Fennec from Fdroid allows about:config and also it's based on Firefox stable, not beta or nightly. Update comes a week or so later than Firefox. But this could be used for this? I haven't tested yet..

I'm using this on Fennec, it works on  Mull and SmartCookieWeb, Waterfox.

**Comment by iHarryPotter178** [comment]
> > Fennec from Fdroid allows about:config and also it's based on Firefox stable, not beta or nightly. Update comes a week or so later than Firefox. But this could be used for this? I haven't tested yet..
> 
> I'm using this on Fennec, it works on Mull and SmartCookieWeb, Waterfox.

I just tested it on fennec, and it works. Now we need a adapted mobile version of betterfox.

**Comment by Eula13** [comment]
Steps for rooted phones so they don't have to use adb: Use a root file explorer and navigate to /data/data/org.mozila.fenix/files/mozilla/"your profile name" and place the user.js there. You can then tweak it to your desire, force stop Firefox, and you're done.

**Comment by SyCoREAPER** [comment]
Completely off topic. This guide serves a very clear and singular purpose. 

If you want tonearm brownie points, create a new thread with step by step for root which is more or less the same as Windows, just different file paths but some general location

**Comment by Eula13** [comment]
> Completely off topic. This guide serves a very clear and singular purpose.
> 
> If you want tonearm brownie points, create a new thread with step by step for root which is more or less the same as Windows, just different file paths but some general location

don't want to add clutter to issues, not completely off topic just offers a really simple way if you have a rooted phone.

**Comment by SyCoREAPER** [comment]
> > Completely off topic. This guide serves a very clear and singular purpose.
> > 
> > If you want tonearm brownie points, create a new thread with step by step for root which is more or less the same as Windows, just different file paths but some general location
> 
> don't want to add clutter to issues, not completely off topic just offers a really simple alternative way if you have a rooted phone. 

I understand where you're going with that thought train but your comment will just get lost in the clutter. A new thread and earning credit for it I think is appropriate in the situation.

**Comment by iHarryPotter178** [comment]
> Steps for rooted phones so they don't have to use adb: Use a root file explorer and navigate to /data/data/org.mozila.fenix/files/mozilla/"your profile name" and place the user.js there. You can then tweak it to your desire, force stop Firefox, and you're done.

does it actually work? i tried it in fennec and mull, but can't tell if it's doing anything, even after looking at the about:config..how to verify it's working on?

**Comment by Eula13** [comment]
> does it actually work? i tried it in fennec and mull, but can't tell if it's doing anything, even after looking at the about:config..how to verify it's working on?

I use firefox nightly and verified using about:config by doing a comparison on user js and about:config

**Comment by RaflIsLife** [comment]
> > Steps for rooted phones so they don't have to use adb: Use a root file explorer and navigate to /data/data/org.mozila.fenix/files/mozilla/"your profile name" and place the user.js there. You can then tweak it to your desire, force stop Firefox, and you're done.
> 
> does it actually work? i tried it in fennec and mull, but can't tell if it's doing anything, even after looking at the about:config..how to verify it's working on?

I had the same issue, and managed to fix it. The way to do this is by changing the chmod permission to 777 for user.js

**Comment by FooqX** [comment]
> > > Steps for rooted phones so they don't have to use adb: Use a root file explorer and navigate to /data/data/org.mozila.fenix/files/mozilla/"your profile name" and place the user.js there. You can then tweak it to your desire, force stop Firefox, and you're done.
> > 
> > 
> > does it actually work? i tried it in fennec and mull, but can't tell if it's doing anything, even after looking at the about:config..how to verify it's working on?
> 
> I had the same issue, and managed to fix it. The way to do this is by changing the chmod permission to 777 for user.js

How to do that?

**Comment by Adhjie** [comment]
this needs to be a separate docs after desktop: 'betterfox-android'
Not sure if this method is the same as inputting user_pref("services.sync.prefs.dangerously_allow_arbitrary", true) to android with the same pref set to true too in desktop.

in any case there should be a warning on using this with firefox android (fenix) stable release, since the about:config is not accessible there (I left a review about this in Gplay, and then dev answered it as not on the roadmap/plan, but is being considered), and the only way to set to default is to change desktop pref as default then do the "services.sync.prefs.dangerously_allow_arbitrary" sync again;
Or input a different value since the specs of each device usually are different.

https://searchfox.org/mozilla-central/source/services/sync/modules/engines/prefs.sys.mjs
https://superuser.com/questions/466291/firefox-sync-aboutconfig
https://support.mozilla.org/en-US/kb/sync-custom-preferences

my use case involves setting cache.memory and cache.disk capacity at different values, so that's why I set the sync prefs as false.
or use the method I explain above.

```
user_pref("services.sync.prefs.sync.browser.cache.disk.capacity", false); // personal pref is string in KiB 1GiB for mobile, 10 GiB for desktop. pref.sync here should be false
user_pref("services.sync.prefs.sync-seen.browser.cache.disk.capacity", false); // personal pref is string in KiB 1GiB for mobile, 10 GiB for desktop. pref.sync here should be false
user_pref("services.sync.prefs.sync-seen.services.sync.prefs.sync.browser.cache.disk.capacity", false); // personal pref is string in KiB 1GiB for mobile, 10 GiB for desktop. pref.sync here should be false

[WIP]
//services.sync.prefs.sync.browser.cache.disk.enable	true	
//services.sync.prefs.sync.browser.cache.disk.smart_size.enabled	false	
//services.sync.prefs.sync.browser.cache.disk_cache_ssl	false	
//services.sync.prefs.sync.browser.cache.memory.capacity	false in mobile desktop:true	
//services.sync.prefs.sync.browser.cache.memory.enable	false:mobile true:desktop
```

I agreed, this should be in the docs, though.
although yes, even though mismatching prefs are usually ignored, when firefox android about:config is misconfigured, it's usually very hard to restore it besides reinstalling it. not sure if mismatched prefs are ignored rules are on all platforms, or not. But I think this is generally safe to be done in firefox beta android since about:config page is accessible there.

**Comment by samloyp** [comment]
I followed the instructions stated and got it maybe working on firefox on android (stable). 
If you go to `about:support`, scroll to about middle part, there's a section called **_Important Modified Preferences_** and it lists all the betterfox user.js. 
See screenshot.
So maybe it worked.

<img width="1080" height="1739" alt="Image" src="https://github.com/user-attachments/assets/a6496703-7709-4a92-a185-8f59d5c58794" />

**Comment by Adhjie** [comment]
@furdiburd 
yeah, that url is a gem, found it from an old thread in one of firefox forum websites.

chrome there is older than the trademark chrome, i think its in a dictionary. chrome team just like the term in browser lingo, so it used it as a name.
gecko is totally about the firefox gecko engine.
I havent delve further but these intricacies are even noticeable, in how many names are mentioned in user-agent of browsers eg in browserleaks.com/ip or javascript links.

**Comment by firepainting** [comment]
@furdiburd 
> Just adding that this should work on normal firefox too. You can access about:config by opening this "chrome://geckoview/content/config.xhtml"

(Noob here, please be kind)

Do the changes I make there really register?

As someone else said here, I go to `about:support` but I don't find the changes I made in "Important Modified Preferences" unless `browser.ml.enable` is not "important." This is true for Firefox Stable, Beta, and Nightly.

Otherwise, the changes I made are still in `chrome://geckoview/content/config.xhtml` even after rebooting Firefox (all of them). 

(I'd rather stick to Firefox Stable and not root my device, which I think this thread is for.)
